Split the persistence thread and the message handler.

baldersheim opened the pull request:
- Let FileStorManager own and control the Component and PersistenceHandler
  separately from the Persistence thread.
- Let FileStorManager allocate and control stripe assignment.

@vekterli and @geirst PR

baldersheim committed:
Split the persistence thread and the message handler.
- Let FileStorManager own and control the Component and PersistenceHandler
  separately from the Persistence thread.
- Let FileStorManager allocate and control stripe assignment.

geirst commented on storage/src/tests/persistence/filestorage/filestormanagertest.cpp:
```diff
@@ -721,12 +714,14 @@ TEST_F(FileStorManagerTest, priority) {
-    std::unique_ptr<DiskThread> thread2(createThread(
-            *config, *_node, _node->getPersistenceProvider(),
-            filestorHandler, bucketOwnershipNotifier, *metrics.disks[0]->threads[1]));
+    vespa::config::content::StorFilestorConfig cfg;
```
The setup code per test function is very similar, and we could bundle it together in a helper class. Something to consider for upcoming work.

baldersheim replied:
I have noticed the same, will do on next encounter.

geirst commented on storage/src/vespa/storage/persistence/persistencehandler.cpp:
```diff
@@ -0,0 +1,162 @@
+// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
Please update copyright.

baldersheim replied:
Fixed

geirst commented on storage/src/vespa/storage/persistence/persistencehandler.cpp:
```diff
@@ -0,0 +1,162 @@
+#include <thread>
+
+#include <vespa/log/log.h>
+LOG_SETUP(".persistence.thread");
```
Please update log target.

baldersheim replied:
Fixed

geirst commented on storage/src/vespa/storage/persistence/persistencehandler.cpp:
```diff
@@ -0,0 +1,162 @@
+        case RecheckBucketInfoCommand::ID:
+            return _splitJoinHandler.handleRecheckBucketInfo(static_cast<RecheckBucketInfoCommand&>(msg), std::move(tracker));
+        default:
+            LOG(warning, "Persistence thread received unhandled internal command %s", msg.toString().c_str());
```
Please update message.

baldersheim replied:
Fixed

geirst commented on storage/src/vespa/storage/persistence/persistencehandler.h:
```diff
@@ -0,0 +1,50 @@
+// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
Please update copyright.

baldersheim replied:
Fixed

geirst commented on storage/src/vespa/storage/persistence/persistencehandler.h:
```diff
@@ -0,0 +1,50 @@
+
+class BucketOwnershipNotifier;
+
+class PersistenceHandler : public Types
```
Consider adding class comment.

baldersheim replied:
Done

geirst reviewed (approved):
👍

vekterli commented on storage/src/vespa/storage/persistence/persistencehandler.h:
```diff
@@ -0,0 +1,50 @@
+{
+public:
+    PersistenceHandler(vespalib::ISequencedTaskExecutor &, ServiceLayerComponent & component,
+                      const vespa::config::content::StorFilestorConfig &, spi::PersistenceProvider &,
```
Nit: indentation off by one

baldersheim replied:
Fixed

vekterli reviewed (approved):
:+1:

baldersheim commented on storage/src/vespa/storage/persistence/filestorage/filestormanager.cpp:
```diff
@@ -133,8 +140,14 @@ FileStorManager::configure(std::unique_ptr<vespa::config::content::StorFilestorC
+                                                         *_config, *_provider, *_filestorHandler,
+                                                         *_bucketOwnershipNotifier, *_metrics->disks[0]->threads[j]));
+            _threads.push_back(std::make_unique<PersistenceThread>(*_persistenceHandlers.back(), *_filestorHandler,
+                                                                   j % numStripes, _component));
```
Fixed

baldersheim commented on storage/src/vespa/storage/persistence/persistencehandler.cpp:
```diff
@@ -0,0 +1,162 @@
+namespace storage {
+
+PersistenceHandler::PersistenceHandler(vespalib::ISequencedTaskExecutor & sequencedExecutor,
+                                     ServiceLayerComponent & component,
```
Fixed

baldersheim committed:
- Add class comments.
- Reduce code visibility.
- Clean up some unused members.
- Some code unification.
